src/MultiphaseStaggeredStokesBlockFACOperator.cpp

```diff
@@ -0,0 +1,955 @@
+                    }
+
+                }
+            }
+            // Fill in ghost cells. We only want to use values on our current level to fill in ghost cells.
+            performGhostFilling({ un_idx, us_idx }, level_num);
+
+            // Compute the normal extension of the solution at coarse fine interfaces if we are not on the coarsest
+            // level
+            // TODO: 0 here is coarsest level number, we should set this to a variable.
+            d_sc_bdry_op->setPatchDataIndices({ un_idx, us_idx });
+            const IntVector<NDIM>& ratio = level->getRatioToCoarserLevel();
+            for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+            {
+                Pointer<Patch<NDIM>> patch = level->getPatch(p());
+                // TODO: 1 here is the ghost width to fill, this should be variable.
+                const IntVector<NDIM>& gcw_to_fill = 1;
+                d_sc_bdry_op->computeNormalExtension(*patch, ratio, gcw_to_fill);
+            }
+        }
+        else if (sweep > 0)
+        {
+                performGhostFilling({ un_idx, us_idx }, level_num);
+        }
+
+        // loop through all patches on this level
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            Pointer<Patch<NDIM>> patch = level->getPatch(p());
+            Pointer<CartesianPatchGeometry<NDIM>> pgeom = patch->getPatchGeometry();
+            const double* const dx = pgeom->getDx(); // dx[0] -> x, dx[1] -> y
+            Pointer<CellData<NDIM, double>> thn_data = patch->getPatchData(thn_idx);
+            Pointer<SideData<NDIM, double>> un_data = patch->getPatchData(un_idx);
+            Pointer<SideData<NDIM, double>> us_data = patch->getPatchData(us_idx);
+            Pointer<SideData<NDIM, double>> f_un_data = patch->getPatchData(f_un_idx);
+            Pointer<SideData<NDIM, double>> f_us_data = patch->getPatchData(f_us_idx);
+            Pointer<SideData<NDIM, int>> mask_data = patch->getPatchData(d_mask_idx);
+
+            // Enforce any Dirichlet boundary conditions.
+            if (d_bc_un_helper->patchTouchesDirichletBoundary(patch))
+            {
+                d_bc_un_helper->copyDataAtDirichletBoundaries(un_data, f_un_data, patch);
+            }
+            if (d_bc_us_helper->patchTouchesDirichletBoundary(patch))
+            {
+                d_bc_us_helper->copyDataAtDirichletBoundaries(us_data, f_us_data, patch);
+            }
+
+            if (d_params.isVariableDrag())
+            {
+                // Place var xi implementation here
+                Pointer<SideData<NDIM, double>> xi_data = patch->getPatchData(d_params.xi_idx);
+
+            }// end variable drag
+            else // use constant xi coefficient
+            {
+                // Gauss-Seidel iterations with successive under relaxation
+                for (SAMRAI::pdat::CellIterator<NDIM> ci(patch->getBox()); ci; ci++) // cell-centers
```

Comment:
This is the cause for degraded performance with coarse-fine interfaces. By looping over cell centers and only updating the lower x and y sides of the cell, the upper sides of the cell on the right-most and top-most cells of a box are unchanged. This needs to be replaced with a loop over cell sides.

This does not cause problems for uniform grids because we use periodic conditions, so the DOF's on the top and right most sides are not "true" DOFs.

Reply:
Oh! I should be able to update that tonight, unless you're already working on fixing that then feel free to push your stuff

Reply:
I might have time to do it tomorrow, but if you are able, go ahead.

Reply:
Just pushed the changes for that. Preliminary testing shows 2nd order convergence when max_levels=2.